Add flags to the core queue interface for device-side ring buf/queue descriptor allocation

atgutier opened the pull request:
Add flags to the queue interface to choose whether or not to allocate it in device memory.

Now the queue struct is also allocated in device memory.

Fixes a bug due to missing the executable flag when allocating the ring buffer in device memory.

atgutier committed:
rocr: Remove empty shared.cpp

atgutier committed:
rocr: Flags to alloc queue buf/struct in dev mem

This builds on a prior change that allowed for allocating
a user-mode queue's packet buffer in device memory to also
allocate the queue struct in device memory. This provides
additional latency benefits particularly for cases where
dispatches are performed from the GPU itself. Flags are
added to support the various use cases.

atgutier commented:
FYI @benvanik this is still WIP, but let me know if this works for you and if the cacheable flag works/has benefits.

dayatsin-amd commented on runtime/hsa-runtime/core/common/hsa_table_interface.cpp:
```diff
@@ -1164,6 +1164,18 @@ hsa_status_t HSA_API hsa_amd_queue_set_priority(hsa_queue_t* queue,
 }
 
+// Mirrors Amd Extension Apis
+hsa_status_t HSA_API hsa_amd_queue_create(hsa_agent_t agent, uint32_t size, hsa_queue_type32_t type,
```
Can we use flags instead of the two bool's for dev_mem and cacheable. That way we can add other flags later without adding a new API.

typedef enum {
   HSA_AMD_QUEUE_FLAG_DEVICE_MEM = (1 << 0),
   HSA_AMD_QUEUE_FLAG_CACHEABLE = (1 << 1),
} hsa_amd_queue_flag_t;


hsa_status_t HSA_API hsa_amd_queue_create(hsa_agent_t agent, uint32_t size, hsa_queue_type32_t type,
                                          void (*callback)(hsa_status_t status, hsa_queue_t* source,
                                                           void* data),
                                          void* data, uint32_t private_segment_size,
                                          uint32_t group_segment_size, uint64_t flags,
                                          hsa_queue_t** queue) ;

atgutier replied:
Yeah, great idea.

dayatsin-amd commented on runtime/hsa-runtime/core/runtime/hsa_api_trace.cpp:
```diff
@@ -87,7 +87,7 @@ void HsaApiTable::Init() {
 
   constexpr size_t expected_core_api_table_size = 1016;
-  constexpr size_t expected_amd_ext_table_size = 584;
+  constexpr size_t expected_amd_ext_table_size = 592;
```
Need to bump HSA_AMD_EXT_API_TABLE_STEP_VERSION

atgutier replied:
Done

dayatsin-amd commented on runtime/hsa-runtime/inc/hsa_ext_amd.h:
```diff
@@ -2813,6 +2813,93 @@ typedef enum hsa_amd_queue_priority_s {
 hsa_status_t HSA_API hsa_amd_queue_set_priority(hsa_queue_t* queue,
                                                 hsa_amd_queue_priority_t priority);
 
+/**
```
Need to bump HSA_AMD_INTERFACE_VERSION_MINOR

atgutier replied:
Done

jayhawk-commits commented:
/AzurePipelines run

atgutier commented on runtime/hsa-runtime/core/runtime/amd_blit_kernel.cpp:
```diff
@@ -1272,7 +1272,7 @@ void BlitKernel::PopulateQueue(uint64_t index, uint64_t code_handle, void* args,
   queue_buffer[index & queue_bitmask_] = packet;
   std::atomic_thread_fence(std::memory_order_release);
-  if (core::Runtime::runtime_singleton_->flag().dev_mem_queue() && !queue_->needsPcieOrdering()) {
+  if (queue_->IsDeviceMem() && !queue_->needsPcieOrdering()) {
```
@saleelk I'm curious if the logic is correct for the original needsPcieOrdering() method you added. Shouldn't this really be:

`queue_->needsPcieOrdering()`

Meaning we need to change the logic of that call internally?

dayatsin-amd reviewed (approved):
Looks good to me. Thank you!

saleelk commented on runtime/hsa-runtime/inc/hsa_ext_amd.h:
```diff
@@ -2828,6 +2829,102 @@ typedef enum hsa_amd_queue_priority_s {
+   * The queue packet buffer and the queue struct should be allocated in
+   * the agent's device memory.
+   */
+  HSA_AMD_QUEUE_FLAG_DEVICE_MEM = (1 << 0),
```
We should explicitly mention whether its cached on uncached. Like HSA_AMD_QUEUE_HOST = 0, HSA_AMD_QUEUE_DEV_UNCACHED = 1 << 0,

atgutier replied:
Removed the cache flag for now.

saleelk commented on runtime/hsa-runtime/inc/hsa_ext_amd.h:
```diff
@@ -2828,6 +2829,102 @@ typedef enum hsa_amd_queue_priority_s {
+   * Used to indicate if the queue created in device memory should be
+   * cacheable.
+   */
+  HSA_AMD_QUEUE_FLAG_CACHEABLE = (1 << 1),
```
We should probably discuss if this is even doable at the moment, to flush L2 is costly and if we are writing a packet (aka 64bytes), we can probably also have false sharing issues/

atgutier replied:
Removed

saleelk reviewed (commented):
the change loosk good except for the flags I mentioned and if we should expose them, and thanks for fixing the typo I had.

atgutier committed:
rocr/libhsakmt: Add coarse-grain allocator to GPU

atgutier committed:
rocr: Add large_bar_enabled var to the GPU agent

Adds a bool to the GPU agent and a public member method to
check if the GPU supports large BAR. This is needed so we can
check if large BAR is supported when a user tries to allocate
an AQL queue in device memory on a given GPU agent.

Also adds an exception to the AQL queue if device-side AQL queues
are requested and the GPU owner of the AQL doesn't support large
BAR. Otherwise, ROCr will currently allow device-side queues
that can cause faults when the user tries to touch their ring
buffers and the user will not know why the faults are occuring.

This relies on the fact that the KFD does not exposed any links
from the CPU to the GPU if large BAR is not enabled (though
links from the GPU to the CPU may still be exposed by the KFD).

misos1 commented:
I was going to test this but I noticed there is no longer any such function as `hsa_amd_queue_create`, only `hsa_queue_create` which sets just one of the new flags based on ENV.

atgutier commented:
> I was going to test this but I noticed there is no longer any such function as `hsa_amd_queue_create`, only `hsa_queue_create` which sets just one of the new flags based on ENV.

We're still fleshing out an API definition to expose this. I will re-post the API portion of the original PR as a WIP tomorrow.

atgutier commented:
I've made all the necessary changes. Could I get another review @dayatsin-amd

atgutier commented:
> I was going to test this but I noticed there is no longer any such function as `hsa_amd_queue_create`, only `hsa_queue_create` which sets just one of the new flags based on ENV.

https://github.com/atgutier/ROCR-Runtime/pull/1

You can use this if you want to play around with this feature, but this is a WIP and should not be relied upon for more than experimentation.

dayatsin-amd commented:
!verify

dayatsin-amd commented:
!verify

dayatsin-amd committed:
Merge branch 'amd-staging' into atgutier/queue-struct-dev-mem

dayatsin-amd commented:
!verify

dayatsin-amd commented:
!verify

dayatsin-amd commented:
@ammallya Can you please check why I cannot import this PR into the github-emu?

dayatsin-amd commented:
Manually pulling in the changes:
https://github.com/AMD-ROCm-Internal/ROCR-Runtime/pull/138
